token-metadata/interface/src/instruction.rs

```diff
@@ -384,4 +396,104 @@ mod test {
+        let serialized_expected =
+            "{\"Initialize\":{\"name\":\"Token Name\",\"symbol\":\"TST\",\"uri\":\"uri.test\"}}";
+        assert_eq!(&serialized, serialized_expected);
+
+        let deserialized = serde_json::from_str::<TokenMetadataInstruction>(&serialized).unwrap();
+        assert_eq!(ix, deserialized);
+    }
+
+    #[cfg(feature = "serde-traits")]
+    #[test]
+    fn update_field_serde() {
+        let data = UpdateField {
+            field: Field::Key("MyField".to_string()),
+            value: "my field value".to_string(),
+        };
+        let ix = TokenMetadataInstruction::UpdateField(data);
+        let serialized = serde_json::to_string(&ix).unwrap();
+        let serialized_expected =
+            "{\"UpdateField\":{\"field\":{\"Key\":\"MyField\"},\"value\":\"my field value\"}}";
+        assert_eq!(&serialized, serialized_expected);
+
+        let deserialized = serde_json::from_str::<TokenMetadataInstruction>(&serialized).unwrap();
+        assert_eq!(ix, deserialized);
+    }
+
+    #[cfg(feature = "serde-traits")]
+    #[test]
+    fn remove_key_serde() {
+        let data = RemoveKey {
+            key: "MyTestField".to_string(),
+            idempotent: true,
+        };
+        let ix = TokenMetadataInstruction::RemoveKey(data);
+        let serialized = serde_json::to_string(&ix).unwrap();
+        let serialized_expected = "{\"RemoveKey\":{\"idempotent\":true,\"key\":\"MyTestField\"}}";
+        assert_eq!(&serialized, serialized_expected);
+
+        let deserialized = serde_json::from_str::<TokenMetadataInstruction>(&serialized).unwrap();
+        assert_eq!(ix, deserialized);
+    }
+
+    #[cfg(feature = "serde-traits")]
+    #[test]
+    fn update_authority_serde() {
+        let update_authority_option: Option<Pubkey> =
+            Some(Pubkey::from_str("4uQeVj5tqViQh7yWWGStvkEG1Zmhx6uasJtWCJziofM").unwrap());
+        let update_authority: OptionalNonZeroPubkey = update_authority_option.try_into().unwrap();
+        let data = UpdateAuthority {
+            new_authority: update_authority,
+        };
+        let ix = TokenMetadataInstruction::UpdateAuthority(data);
+        let serialized = serde_json::to_string(&ix).unwrap();
+        let serialized_expected = "{\"UpdateAuthority\":{\"new_authority\":\"4uQeVj5tqViQh7yWWGStvkEG1Zmhx6uasJtWCJziofM\"}}";
+        assert_eq!(&serialized, serialized_expected);
+
+        let deserialized = serde_json::from_str::<TokenMetadataInstruction>(&serialized).unwrap();
+        assert_eq!(ix, deserialized);
+    }
+
+    #[cfg(feature = "serde-traits")]
+    #[test]
+    fn update_authority_serde_with_none() {
+        let data = UpdateAuthority {
+            new_authority: OptionalNonZeroPubkey::default(),
+        };
+        let ix = TokenMetadataInstruction::UpdateAuthority(data);
+        let serialized = serde_json::to_string(&ix).unwrap();
+        let serialized_expected = "{\"UpdateAuthority\":{\"new_authority\":null}}";
```

Comment:
I'm just now realizing that we did these a bit wrong... since it's meant to be consumed by JSON, we should really use JS idioms and do everything `camelCase`. We'll fix token-2022 separately, but for this one, can you make everything camelCase? Ie. add another line like:

```
#[cfg_attr(feature = "serde-traits", serde(rename_all = "camelCase"))]
```

More info at https://serde.rs/attr-rename.html

Reply:
Sure thing.

The attributes we're adding for `serde` are quite a few now. I was thinking maybe it's worth investing time in creating a `proc_macro_attribute` that will automatically add the appropriate attributes depending on the type.

So we would go from this:

```
#[cfg_attr(feature = "serde-traits", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "serde-traits", serde(rename_all = "camelCase"))]
#[derive(Clone, Copy, Debug, PartialEq)]
#[repr(u8)]
pub enum TransferFeeInstruction {
    InitializeTransferFeeConfig {
        #[cfg_attr(feature = "serde-traits", serde(with = "coption_fromstr"))]
        transfer_fee_config_authority: COption<Pubkey>,
        #[cfg_attr(feature = "serde-traits", serde(with = "coption_fromstr"))]
        withdraw_withheld_authority: COption<Pubkey>,
        transfer_fee_basis_points: u16,
        maximum_fee: u64,
    },
...
}

#[cfg_attr(feature = "serde-traits", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "serde-traits", serde(from = "u64", into = "u64"))]
#[derive(Clone, Copy, Debug, Default, PartialEq, Pod, Zeroable)]
#[repr(transparent)]
pub struct PodU64([u8; 8]);

```

To something like this:

```
#[cfg_attr(feature = "serde-traits", spl_serde)]
#[derive(Clone, Copy, Debug, PartialEq)]
#[repr(u8)]
pub enum TransferFeeInstruction {
    InitializeTransferFeeConfig {
        transfer_fee_config_authority: COption<Pubkey>,
        withdraw_withheld_authority: COption<Pubkey>,
        transfer_fee_basis_points: u16,
        maximum_fee: u64,
    },
...
}

#[cfg_attr(feature = "serde-traits", spl_serde)]
#[derive(Clone, Copy, Debug, Default, PartialEq, Pod, Zeroable)]
#[repr(transparent)]
pub struct PodU64([u8; 8]);
```

Since `serde` logic is spread out in a few crates (`program-2022` and `token-metadata` now) this `proc_macro_attribute` would reside in `/libraries` so it can be used in several crates.

Do you think this is worth it? I could implement it.

Reply:
I think that makes sense. Maybe you can work with @buffalojoec to create a little crate with the macro? For this PR, let's stick with the two lines

Reply:
Done. 

Will sync with @buffalojoec regarding the proc_macro_attribute.